### PR TITLE
[ci][postmerge/1] create a scheduler for postmerge pipeline

### PR DESCRIPTION
**Pull request by can-anyscale**
Create a scheduler that implements the fill-in-the-gap for postmerge pipeline. This PR:

- create a skeleton class with buildkite credentials
- a function to get the latest build given a status and some test cases

Test:
- CI

### CODE REVIEW

**Review by aslonnie — COMMENTED** [review]
where do we plan to run this scheduler?

**Comment by can-anyscale** [comment]
the plan is to run this as a buildkite job yeah

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            [build for build in self._get_builds() if build["state"] == build_state],
+            key=lambda build: build["number"],
+            reverse=True,
+        )
+        if not builds:
+            return None
+
+        return builds[0]["commit"]
+
+    def _get_builds(self, days_ago: int = 1) -> List[Dict[str, Any]]:
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/gap_filling_scheduler.py:44]: can we drop the default `days_ago = 1` here, but make it a config field in the scheduler?

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        return self.buildkite.builds().list_all_for_pipeline(
+            self.buildkite_organization,
+            self.buildkite_pipeline,
+            created_from=datetime.now() - timedelta(days=days_ago),
+            branch="master",
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/gap_filling_scheduler.py:49]: why it has to be master branch? should this also be configurable?

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        self.buildkite_pipeline = buildkite_pipeline
+        GapFillingScheduler._init_buildkite()
+
+    @classmethod
+    def _init_buildkite(cls):
+        if not cls.buildkite:
+            cls.buildkite = Buildkite()
+            cls.buildkite.set_access_token(get_secret_token(AWS_SECRET_BUILDKITE))
+
+    def _get_latest_revision_for_build_state(self, build_state: str) -> Optional[str]:
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/gap_filling_scheduler.py:33]: 1. latest "commit" rather than "revision"?
2. what if there is a manual build on a commit from the past on the pipeline? like for recreating a wheel or re-testing something? I think to get the recent history of the branch and cross reference with it to get the correct logical order of the commits, rather than relying buildkite's build number.

File: ci/ray_ci/BUILD.bazel [hunk 4]
```diff
@@ -4,9 +4,10 @@ load("@py_deps_buildkite//:requirements.bzl", ci_require = "requirement")
 py_library(
     name = "ray_ci_lib",
     srcs = glob(
-        ["*.py"],
+        ["*.py", "pipeline/*.py"],
```
[the remark this answers is not in the document]
Comment (can-anyscale) [reply ci/ray_ci/BUILD.bazel:7]: sound good

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+from datetime import datetime, timedelta
+from typing import List, Dict, Optional, Any
+
+from pybuildkite.buildkite import Buildkite
+
+from release.ray_release.aws import get_secret_token
+from release.ray_release.configs.global_config import AWS_SECRET_BUILDKITE
```
[the remark this answers is not in the document]
Comment (can-anyscale) [reply ci/ray_ci/pipeline/gap_filling_scheduler.py:7]: if we move can we move the release + ci/ray_ci all together; i think having them in one repo is better; when the time comes to modulize the release package; we can have a core/common module for the global config that can be shared across other modules

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    ):
+        self.buildkite_organization = buildkite_organization
+        self.buildkite_pipeline = buildkite_pipeline
+        GapFillingScheduler._init_buildkite()
+
+    @classmethod
+    def _init_buildkite(cls):
+        if not cls.buildkite:
+            cls.buildkite = Buildkite()
+            cls.buildkite.set_access_token(get_secret_token(AWS_SECRET_BUILDKITE))
```
[the remark this answers is not in the document]
Comment (can-anyscale) [reply ci/ray_ci/pipeline/gap_filling_scheduler.py:31]: i want to avoid loading secret token multiple times, but i guess it's not not too bad; the plan is to move the secret token to the global_config (because it is defined once per repository, not the GappFillingScheduler class)

File: ci/ray_ci/pipeline/gap_filling_scheduler.py (hunk 2 above)
Comment (can-anyscale) [reply ci/ray_ci/pipeline/gap_filling_scheduler.py:49]: let's start with what we need first (for postmerge); we can make it more flexible when the need arises

File: ci/ray_ci/pipeline/gap_filling_scheduler.py (hunk 3 above)
Comment (can-anyscale) [reply ci/ray_ci/pipeline/gap_filling_scheduler.py:33]: good point; the logic is a bit involved so let me put a TODO to make the logic more precise in subsequent PR, to avoid bloating this PR

File: ci/ray_ci/pipeline/gap_filling_scheduler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aslonnie) [reply ci/ray_ci/pipeline/gap_filling_scheduler.py:31]: you can always send in the config as a arg or parameter. global configs should not be imported into sub modules, but should be dispatched into sub modules on init process.

otherwise, you would have a config file ever growing, with business logic from all modules, all mixed together, and have a dependency structure that every logic depends on every other logic's config structs.

File: ci/ray_ci/pipeline/gap_filling_scheduler.py (hunk 3 above)
Comment (aslonnie) [reply ci/ray_ci/pipeline/gap_filling_scheduler.py:33]: I worry that this logic is just wrong for these manually triggered corner cases, and can lead to a lot of unnecessary bisect runs when being used.

to allow you moving forward.. maybe for this PR, rather than picking one, just return the entire list of recent history?

Comment (can-anyscale) [reply ci/ray_ci/pipeline/gap_filling_scheduler.py:33]: yes, i won't run this in production, just want to break it down for easier of review; i can also see how much additional logic is added, if not to much i can just add it here

File: ci/ray_ci/pipeline/gap_filling_scheduler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (can-anyscale) [reply ci/ray_ci/pipeline/gap_filling_scheduler.py:31]: sound good, i'll add the global configs only on the top level script and pass down only the required values here

**Comment by can-anyscale** [comment]
@aslonnie's comments

File: ci/ray_ci/pipeline/BUILD.bazel [hunk 7]
```diff
@@ -0,0 +1,31 @@
+load("@rules_python//python:defs.bzl", "py_library", "py_test")
+load("@py_deps_buildkite//:requirements.bzl", ci_require = "requirement")
+
+py_library(
+    name = "ray_ci_pipeline_lib",
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/BUILD.bazel:5]: you can just call this library `pipeline`. it is common for the main build target to have the same name as the dir/package name.

File: ci/ray_ci/pipeline/test_gap_filling_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+from unittest import mock
+
+import pytest
+
+from ci.ray_ci.pipeline.gap_filling_scheduler import GapFillingScheduler
+
+
+@mock.patch(
+    "ci.ray_ci.pipeline.gap_filling_scheduler.GapFillingScheduler._get_latest_commits"
+)
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/test_gap_filling_scheduler.py:11]: rather than mocking this, can we get a temp work dir, and create a real git repo for test there with fake history?

I prefer using real git repo for all git related operations and testing.

File: ci/ray_ci/pipeline/test_gap_filling_scheduler.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+            "commit": "222",
+        },
+    ]
+    mock_get_latest_commits.return_value = [
+        "222",
+        "111",
+        "000",
+    ]
+    scheduler = GapFillingScheduler("org", "pipeline", "token")
+    assert scheduler._get_latest_commit_for_build_state("failed") == "222"
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/test_gap_filling_scheduler.py:34]: add test cases there recent builds or recent commits are empty?

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+    def _get_latest_commit_for_build_state(self, build_state: str) -> Optional[str]:
+        latest_commits = self._get_latest_commits()
+        builds = sorted(
+            [
+                build
+                for build in self._get_builds()
+                if build["state"] == build_state and build["commit"] in latest_commits
+            ],
+            key=lambda build: latest_commits.index(build["commit"]),
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/gap_filling_scheduler.py:38]: build a commit->index dict? so that we do not do this `O(MxN)` lookup?

also, can we have some tie breaking mechanism for builds that run on the same commit? maybe based on the build's number?

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        )
+        if not builds:
+            return None
+
+        return builds[0]["commit"]
+
+    def _get_latest_commits(self) -> List[str]:
+        return (
+            subprocess.check_output(
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/gap_filling_scheduler.py:47]: maybe send in a work dir here.

- bazel run does not run in the dir of the workspace. it can read the location though, so you either needs to send in the dir, or needs to change the work dir
- this would need the dir if we want to handle more than one repos in the same python program.
- if we use a real git repo for testing, then that dir needs to be send in / set up here too.

File: ci/ray_ci/pipeline/gap_filling_scheduler.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        self.buildkite_organization = buildkite_organization
+        self.buildkite_pipeline = buildkite_pipeline
+        self.buildkite = Buildkite()
+        self.buildkite.set_access_token(buildkite_access_token)
+        self.days_ago = days_ago
+
+    def _get_latest_commit_for_build_state(self, build_state: str) -> Optional[str]:
+        latest_commits = self._get_latest_commits()
+        builds = sorted(
+            [
```
Comment (aslonnie) [inline ci/ray_ci/pipeline/gap_filling_scheduler.py:33]: maybe convert this to a more conventional loop? I always find functional language style comprehension hard to read.. maybe it is just that my math is not that good..

File: ci/ray_ci/pipeline/gap_filling_scheduler.py (hunk 10 above)
Comment (can-anyscale) [reply ci/ray_ci/pipeline/gap_filling_scheduler.py:38]: the tie breaking is not required here yeah since we just returns the commit not the build number so any build would work

File: ci/ray_ci/pipeline/gap_filling_scheduler.py (hunk 11 above)
Comment (can-anyscale) [reply ci/ray_ci/pipeline/gap_filling_scheduler.py:47]: ah yes, forgot about cwd

File: ci/ray_ci/pipeline/test_gap_filling_scheduler.py (hunk 8 above)
Comment (can-anyscale) [reply ci/ray_ci/pipeline/test_gap_filling_scheduler.py:11]: do you want an e2e test? i think unit-test should mock the inputs; i can write a buildkite job as a e2e test once i have a binary to execute this e2e if you want

**Review by aslonnie — APPROVED** [review]
